pytensor/tensor/rewriting/linalg.py

```diff
@@ -539,3 +540,54 @@ def svd_uv_merge(fgraph, node):
                     or len(fgraph.clients[cl.outputs[2]]) > 0
                 ):
                     return [cl.outputs[1]]
+
+
+def _find_solve_with_eye(node):
+    valid_solves = (Solve, SolveTriangular)
+    # First, we look for the solve op
+    if not (
+        isinstance(node.op, Blockwise) and isinstance(node.op.core_op, valid_solves)
+    ):
+        return False
+    # Check whether second input to solve is Eye
+    solve_inputs = node.inputs
+    potential_eye_input = solve_inputs[1].owner
+    if not (potential_eye_input and isinstance(potential_eye_input.op, Eye)):
+        return False
+
+    return True
+
+
+@register_canonicalize
+@register_stabilize
+@node_rewriter([Blockwise])
+def rewrite_inv_inv(fgraph, node):
+    valid_inverses = (MatrixInverse, MatrixPinv, Solve, SolveTriangular)
+    valid_solves = (Solve, SolveTriangular)
+    # Check if Solve has b = eye
+    solve_inv_check = True
+    if isinstance(node.op.core_op, valid_solves):
+        solve_inv_check = _find_solve_with_eye(node)
```

Comment:
Do we expect this to show up in real cases?

Reply:
yea? we arent actually rewriting solve(a, eye) to inv(a), just identifying if its a nested solve

Reply:
It's widely known that `solve(x, eye)` is more numerically stable than `inv(x)`, so I expect most users will write inverse that way.

Reply:
Cool didn't know that!

Reply:
We still need to probably prevent eye from being constant folded, and perhaps "uncanonicalize" eye constants to the eye operator. Also always need to check that sneaky k=0 case, so would be good to have that as a helper?

To stop eye from constant_folding when it's inputs are constants we need to define the `do_constant_fold` method to return None. I think we should do this to most "tensor" creation ops like eye, alloc, diagonal, arange..., and only constant fold them at a later stage? Gonna convert this comment to an issue